File: frontend/lib/models/inventory.dart
class Product {
  final int id;
  final String name;
  final String sku;
  final String? description;
  final String unit;
  final double costPrice;
  final double sellPrice;
  final String? category;

  Product({
    required this.id,
    required this.name,
    required this.sku,
    this.description,
    required this.unit,
    required this.costPrice,
    required this.sellPrice,
    this.category,
  });

  factory Product.fromJson(Map<String, dynamic> json) {
    return Product(
      id: json['id'],
      name: json['name'],
      sku: json['sku'],
      description: json['description'],
      unit: json['unit'],
      costPrice: (json['cost_price'] as num).toDouble(),
      sellPrice: (json['sell_price'] as num).toDouble(),
      category: json['category'],
    );
  }

  Map<String, dynamic> toJson() => {
        'name': name,
        'sku': sku,
        'description': description,
        'unit': unit,
        'cost_price': costPrice,
        'sell_price': sellPrice,
        'category': category,
      };
}

class Warehouse {
  final int id;
  final String name;
  final String? location;

  Warehouse({required this.id, required this.name, this.location});

  factory Warehouse.fromJson(Map<String, dynamic> json) {
    return Warehouse(
      id: json['id'],
      name: json['name'],
      location: json['location'],
    );
  }
}

class InventoryItem {
  final int id;
  final int productId;
  final int warehouseId;
  final double quantity;

  InventoryItem({
    required this.id,
    required this.productId,
    required this.warehouseId,
    required this.quantity,
  });

  factory InventoryItem.fromJson(Map<String, dynamic> json) {
    return InventoryItem(
      id: json['id'],
      productId: json['product_id'],
      warehouseId: json['warehouse_id'],
      quantity: (json['quantity'] as num).toDouble(),
    );
  }
}

File: app/models/inventory.py
from datetime import datetime, timezone

from sqlalchemy import DateTime, ForeignKey, Integer, Numeric, String, Text
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.db.base import Base


class Product(Base):
    __tablename__ = "products"

    id: Mapped[int] = mapped_column(Integer, primary_key=True, index=True)
    name: Mapped[str] = mapped_column(String(200), nullable=False)
    sku: Mapped[str] = mapped_column(String(100), unique=True, nullable=False, index=True)
    description: Mapped[str | None] = mapped_column(Text)
    unit: Mapped[str] = mapped_column(String(50), default="piece")
    cost_price: Mapped[float] = mapped_column(Numeric(12, 2), nullable=False)
    sell_price: Mapped[float] = mapped_column(Numeric(12, 2), nullable=False)
    category: Mapped[str | None] = mapped_column(String(100))
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), default=lambda: datetime.now(timezone.utc))


class Warehouse(Base):
    __tablename__ = "warehouses"

    id: Mapped[int] = mapped_column(Integer, primary_key=True, index=True)
    name: Mapped[str] = mapped_column(String(200), nullable=False, unique=True)
    location: Mapped[str | None] = mapped_column(String(300))

    inventory: Mapped[list["Inventory"]] = relationship(back_populates="warehouse")


class Inventory(Base):
    __tablename__ = "inventory"

    id: Mapped[int] = mapped_column(Integer, primary_key=True, index=True)
    product_id: Mapped[int] = mapped_column(Integer, ForeignKey("products.id"), nullable=False)
    warehouse_id: Mapped[int] = mapped_column(Integer, ForeignKey("warehouses.id"), nullable=False)
    quantity: Mapped[float] = mapped_column(Numeric(12, 2), default=0)

    warehouse: Mapped["Warehouse"] = relationship(back_populates="inventory")


class InventoryMovement(Base):
    __tablename__ = "inventory_movements"

    id: Mapped[int] = mapped_column(Integer, primary_key=True, index=True)
    product_id: Mapped[int] = mapped_column(Integer, ForeignKey("products.id"), nullable=False)
    warehouse_id: Mapped[int] = mapped_column(Integer, ForeignKey("warehouses.id"), nullable=False)
    movement_type: Mapped[str] = mapped_column(String(20), nullable=False)
    quantity: Mapped[float] = mapped_column(Numeric(12, 2), nullable=False)
    reference: Mapped[str | None] = mapped_column(String(200))
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), default=lambda: datetime.now(timezone.utc))
